spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+
+  def makeRelation(
+      table: String,
+      keyspace: String,
+      cluster: Option[String],
+      userSpecifiedSchema: Option[StructType],
+      connector: CassandraConnector,
+      readConf: ReadConf,
+      writeConf: WriteConf,
+      sqlContext: SQLContext) : BaseRelationImpl
+}
+
+/** Base table scan type */
+case object BaseScanType extends ScanType  {
+
+  override def makeRelation(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               userSpecifiedSchema: Option[StructType],
+               connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+               sqlContext: SQLContext) : BaseRelationImpl = {
+
+    new TableScanRelationImpl(
+        table,
+        keyspace,
+        cluster,
+        connector,
+        readConf,
+        writeConf,
+        userSpecifiedSchema,
+        sqlContext)
+  }
+
+}
+
+/** Pruned columns table scan type */
+case object PrunedScanType extends ScanType {
+
+  override def makeRelation(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               userSpecifiedSchema: Option[StructType],
+               connector: CassandraConnector,
+               readConf: ReadConf,
+               writeConf: WriteConf,
+               sqlContext: SQLContext) : BaseRelationImpl = {
+
+    new PrunedScanRelationImpl(
+        table,
+        keyspace,
+        cluster,
+        connector,
+        readConf,
+        writeConf,
+        userSpecifiedSchema,
+        sqlContext)
+  }
+
+}
+
+/** Pruned columns and filtered table scan type */
+case object PrunedFilteredScanType extends ScanType {
+
+  override def makeRelation(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               userSpecifiedSchema: Option[StructType],
+               connector: CassandraConnector,
+               readConf: ReadConf,
+               writeConf: WriteConf,
+               sqlContext: SQLContext) : BaseRelationImpl = {
+
+    new PrunedFilteredScanRelationImpl(
+        table,
+        keyspace,
+        cluster,
+        connector,
+        readConf,
+        writeConf,
+        userSpecifiedSchema,
+        sqlContext)
+  }
+
+}
+
+/** Pruned columns and Catalyst filtered table scan type */
+case object CatalystScanType extends ScanType {
+
+  override def makeRelation(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               userSpecifiedSchema: Option[StructType],
+               connector: CassandraConnector,
+               readConf: ReadConf,
+               writeConf: WriteConf,
+               sqlContext: SQLContext) : BaseRelationImpl = {
+
+    new CatalystScanRelationImpl(
+        table,
+        keyspace,
+        cluster,
+        connector,
+        readConf,
+        writeConf,
+        userSpecifiedSchema,
+        sqlContext)
```

Comment:
For long lists of arguments, it is preferred to use named arguments, in order to avoid messing up the order.
Of course, using short list of arguments is even more preferred.